src/datasets.jl

```diff
@@ -482,6 +482,57 @@ write_dataset(
     xfer::DatasetTransferProperties=dataset.xfer
 ) = nothing
 
+"""
+    function write_compound(f, name, data::AbstractArray{T}) where T
+
+Writes an array of structs with primitive field types to an HDF5Compound dataset.
+
+Examples
+========
+
+```
+julia> struct Foo
+           x::Int32
+           y::Float64
+       end
+
+julia> foos = [[Foo(1, 2) Foo(3, 4) Foo(5, 6)]; [Foo(7, 8) Foo(9, 10) Foo(11, 12)]]
+2×3 Matrix{Foo}:
+ Foo(1, 2.0)  Foo(3, 4.0)   Foo(5, 6.0)
+ Foo(7, 8.0)  Foo(9, 10.0)  Foo(11, 12.0)
+
+julia> h5open("foo.h5", "w") do h5f
+           write_compound(h5f, "the/foo", ts)
+       end
+
+julia> thefoo = h5open("foo.h5", "r") do file
+           read(file, "the/foo")
+       end
+2×3 Matrix{NamedTuple{(:x, :y), Tuple{Int32, Float64}}}:
+ (x = 1, y = 2.0)  (x = 3, y = 4.0)   (x = 5, y = 6.0)
+ (x = 7, y = 8.0)  (x = 9, y = 10.0)  (x = 11, y = 12.0)
+```
+
+The data can be reinterpreted to the original data type using `reinterpret()`:
+
+```
+julia> reinterpret(Foo, thefoo)
+2×3 reinterpret(Foo, ::Matrix{NamedTuple{(:x, :y), Tuple{Int32, Float64}}}):
+ Foo(1, 2.0)  Foo(3, 4.0)   Foo(5, 6.0)
+ Foo(7, 8.0)  Foo(9, 10.0)  Foo(11, 12.0)
+```
+
+"""
+function write_compound(f, name, data::AbstractArray{T}) where {T}
```

Comment:
Can you add some more type information to this?

Also, does the compound type need to be in an AbstractArray?

Reply:
Yes sure, I'll restrict to `AbstractString`. I thought it makes more sense to write arrays of compound structures. I'll have a look!

Reply:
I am a bit confused. Is it a common use-case to store a single struct in a dataset?

Reply:
It can be, especially if we are writing the data type to an attribute. For example, I encountered one case where we are encapsulating a foreign header as a 1024 byte compound type:
https://github.com/mkitti/FIBSEMHeaders.jl/blob/eded59eac531fe4610edc71453c1ace7fc1039f9/src/FIBSEMHeaders.jl#L26

It seems this assumes we are writing compound datasets to a dataset.

Reply:
OK yes, you are right, I will have a look ;)